### ISSUE TITLE
Normalisation for similarity score

### ISSUE THREAD

**Issue by KarthikeyaKaushik** [omoindrot/tensorflow-triplet-loss#13]
Hey!

Thank you for a well written blog, and a well commented piece of code to go with it. The problem I am working on requires that I compute the similarity between two images. As you suggested in another issue, I created my own pipeline to use images. For the purpose of testing, I used the MNIST dataset again. I am having some trouble understanding what to normalise so that pairwise distances lie between 0 and 1. In other words, two images of the same class must have a similarity of 0, and an image ( from possibly untrained class ) from a dissimlar class have a similarity score of 1. Any suggestions?

**Comment by omoindrot** [comment]
Hi @KarthikeyaKaushik,

I think the easiest way to make sure that distances are normalized is to make sure that all the embeddings are in the L2 sphere.

You can modify the code [here][1]:


```python
with tf.variable_scope('model'):
    # Compute the embeddings with the model
    embeddings = build_model(is_training, images, params)

    # L2 normalize the embeddings so that they lie in the L2 unit sphere
    embeddings = tf.nn.l2_normalize(embeddings, axis=1)
```


If you L2 normalize the embeddings, the distance between two embeddings will always be between 0 and 2.  
So your similarity score could just be the distance divided by 2.

Images of the same class should have a distance of 0 (or at least lower than the margin).  
Images of different classes should have a distance between the margin and 2.



[1]: https://github.com/omoindrot/tensorflow-triplet-loss/blob/ff1b41550de719ee4fb417c3430fd1e33dcc834b/model/model_fn.py#L64-L66

**Comment by xiaomingdaren123** [comment]
@omoindrot 
Hi，omoindrot 
    Thanks for your code,I meet some question, loss value is approximate of margin,i found that the distance is close to 0，I don't know how it was caused.
    Does the output of the network need to be L2 normalized,?what is the role of L2 normalization?How to set the value of  margin，if i don't use L2 normalized？

  thanks
